File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/product/ProductFavouriteViewModel.java
package com.dewonderstruck.apps.ashx0.viewmodel.product;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.dewonderstruck.apps.Config;
import com.dewonderstruck.apps.ashx0.repository.product.ProductRepository;
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData;
import com.dewonderstruck.apps.ashx0.utils.Utils;
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel;
import com.dewonderstruck.apps.ashx0.viewobject.Product;
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource;

import java.util.List;

import javax.inject.Inject;

public class ProductFavouriteViewModel extends PSViewModel {

    //for product favourite list
    private final LiveData<Resource<List<Product>>> productFavouriteData;
    private MutableLiveData<ProductFavouriteViewModel.TmpDataHolder> productFavouriteListObj = new MutableLiveData<>();

    private final LiveData<Resource<Boolean>> nextPageFavouriteLoadingData;
    private MutableLiveData<ProductFavouriteViewModel.TmpDataHolder> nextPageLoadingFavouriteObj = new MutableLiveData<>();

    private final LiveData<Resource<Boolean>> sendFavouritePostData;
    private MutableLiveData<TmpDataHolder> sendFavouriteDataPostObj = new MutableLiveData<>();

    //endregion
    //region Constructor

    @Inject
    public ProductFavouriteViewModel(ProductRepository productRepository) {
        //  product detail List
        productFavouriteData = Transformations.switchMap(productFavouriteListObj, obj -> {
            if (obj == null) {
                return AbsentLiveData.create();
            }
            Utils.psLog("productFavouriteData");
            return productRepository.getFavouriteList(Config.API_KEY, obj.loginUserId, obj.offset);
        });

        nextPageFavouriteLoadingData = Transformations.switchMap(nextPageLoadingFavouriteObj, obj -> {
            if (obj == null) {
                return AbsentLiveData.create();
            }
            Utils.psLog("nextPageFavouriteLoadingData");
            return productRepository.getNextPageFavouriteProductList(obj.loginUserId, obj.offset);
        });

        sendFavouritePostData = Transformations.switchMap(sendFavouriteDataPostObj, obj -> {

            if (obj == null) {
                return AbsentLiveData.create();
            }
            return productRepository.uploadFavouritePostToServer(obj.productId, obj.userId);
        });

    }

    //endregion

    //region Getter And Setter for product detail List
    public void setProductFavouriteListObj(String loginUserId, String offset) {
        if (!isLoading) {
            ProductFavouriteViewModel.TmpDataHolder tmpDataHolder = new ProductFavouriteViewModel.TmpDataHolder();
            tmpDataHolder.loginUserId = loginUserId;
            tmpDataHolder.offset = offset;
            productFavouriteListObj.setValue(tmpDataHolder);

            // start loading
            setLoadingState(true);
        }
    }

    public LiveData<Resource<List<Product>>> getProductFavouriteData() {
        return productFavouriteData;
    }
    //endregion


    //Get Favourite Next Page
    public void setNextPageLoadingFavouriteObj(String offset, String loginUserId) {

        if (!isLoading) {
            ProductFavouriteViewModel.TmpDataHolder tmpDataHolder = new ProductFavouriteViewModel.TmpDataHolder();
            tmpDataHolder.loginUserId = loginUserId;
            tmpDataHolder.offset = offset;
            nextPageLoadingFavouriteObj.setValue(tmpDataHolder);

            // start loading
            setLoadingState(true);
        }
    }

    public LiveData<Resource<Boolean>> getNextPageFavouriteLoadingData() {
        return nextPageFavouriteLoadingData;
    }

    //fav post
    public void setFavouritePostDataObj(String product_id, String userId) {

        if(!isLoading) {
            TmpDataHolder tmpDataHolder = new TmpDataHolder();
            tmpDataHolder.productId = product_id;
            tmpDataHolder.userId = userId;

            sendFavouriteDataPostObj.setValue(tmpDataHolder);

            // start loading
            setLoadingState(true);
        }
    }

    public LiveData<Resource<Boolean>> getFavouritePostData() {
        return sendFavouritePostData;
    }

    static class TmpDataHolder {
        public String productId = "";
        public String userId = "";
        public String shopId = "";
        public String loginUserId = "";
        public String offset = "";
        public String limit = "";
        public Boolean isConnected = false;
    }
    //endregion

}

File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/product/ProductFavouriteViewModel.kt
package com.dewonderstruck.apps.ashx0.viewmodel.product

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.dewonderstruck.apps.Config
import com.dewonderstruck.apps.ashx0.repository.product.ProductRepository
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData
import com.dewonderstruck.apps.ashx0.utils.Utils
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel
import com.dewonderstruck.apps.ashx0.viewobject.Product
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource
import javax.inject.Inject

class ProductFavouriteViewModel @Inject constructor(productRepository: ProductRepository) : PSViewModel() {
    //for product favourite list
    val productFavouriteData: LiveData<Resource<List<Product>>>
    private val productFavouriteListObj = MutableLiveData<TmpDataHolder>()
    val nextPageFavouriteLoadingData: LiveData<Resource<Boolean>>
    private val nextPageLoadingFavouriteObj = MutableLiveData<TmpDataHolder>()
    val favouritePostData: LiveData<Resource<Boolean>>
    private val sendFavouriteDataPostObj = MutableLiveData<TmpDataHolder>()

    //endregion
    //region Getter And Setter for product detail List
    fun setProductFavouriteListObj(loginUserId: String, offset: String) {
        if (!isLoading) {
            val tmpDataHolder = TmpDataHolder()
            tmpDataHolder.loginUserId = loginUserId
            tmpDataHolder.offset = offset
            productFavouriteListObj.value = tmpDataHolder

            // start loading
            setLoadingState(true)
        }
    }

    //endregion
    //Get Favourite Next Page
    fun setNextPageLoadingFavouriteObj(offset: String, loginUserId: String) {
        if (!isLoading) {
            val tmpDataHolder = TmpDataHolder()
            tmpDataHolder.loginUserId = loginUserId
            tmpDataHolder.offset = offset
            nextPageLoadingFavouriteObj.value = tmpDataHolder

            // start loading
            setLoadingState(true)
        }
    }

    //fav post
    fun setFavouritePostDataObj(product_id: String, userId: String) {
        if (!isLoading) {
            val tmpDataHolder = TmpDataHolder()
            tmpDataHolder.productId = product_id
            tmpDataHolder.userId = userId
            sendFavouriteDataPostObj.value = tmpDataHolder

            // start loading
            setLoadingState(true)
        }
    }

    internal class TmpDataHolder {
        var productId = ""
        var userId = ""
        var shopId = ""
        var loginUserId = ""
        var offset = ""
        var limit = ""
        var isConnected = false
    } //endregion

    //endregion
    //region Constructor
    init {
        //  product detail List
        productFavouriteData = Transformations.switchMap(productFavouriteListObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<List<Product>>>()
            }
            Utils.psLog("productFavouriteData")
            productRepository.getFavouriteList(Config.API_KEY, obj.loginUserId, obj.offset)
        }
        nextPageFavouriteLoadingData = Transformations.switchMap(nextPageLoadingFavouriteObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<Boolean>>()
            }
            Utils.psLog("nextPageFavouriteLoadingData")
            productRepository.getNextPageFavouriteProductList(obj.loginUserId, obj.offset)
        }
        favouritePostData = Transformations.switchMap(sendFavouriteDataPostObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<Boolean>>()
            }
            productRepository.uploadFavouritePostToServer(obj.productId, obj.userId)
        }
    }
}